[JIT] Display error message when NamedTuples have user-defined functions

SplitInfinity opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#49115 [JIT] Display error message when NamedTuples have user-defined functions**

**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

Differential Revision: [D25490367](https://our.internmc.facebook.com/intern/diff/D25490367)

Commit:
[JIT] Display error message when NamedTuples have user-defined function

**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

[ghstack-poisoned]

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

[ghstack-poisoned]

SplitInfinity commented on torch/_jit_internal.py:
```diff
@@ -33,6 +33,19 @@
 boolean_dispatched: 'weakref.WeakKeyDictionary[Callable, Dict[str, Callable]]' = weakref.WeakKeyDictionary()  # noqa: T484
 
 
+class FrontendError(Exception):
```
FYI: I moved this in order to get the source range to render properly and consistently.

SplitInfinity commented on torch/_jit_internal.py:
```diff
@@ -216,6 +229,43 @@ def __getattr__(self, key):
+
+    for name, value in cls.__dict__.items():
+        if name not in allowed_methods and inspect.isfunction(value):
+            raise FrontendError(loc, "User-defined methods in NamedTuples are currently not supported")
```
Even this is not quite what we want - the `loc` here is the source location of the use of the name of the `NamedTuple` class that caused it to be compiled, not the source code for it. So, the `loc` doesn't highlight the name or source of the problematic method.

SplitInfinity commented on torch/jit/annotations.py:
```diff
@@ -335,6 +335,8 @@ def try_ann_to_type(ann, loc):
     def fake_rcb(key):
         return None
+    # If ann is a NamedTuple, check that it meet requirements for JIT compilation.
+    torch._jit_internal.check_can_compile_named_tuple(ann, loc)
```
Should this be done in C++ instead?

SplitInfinity replied:
Good point, let me move this to C++ and call it only when we are sure we are dealing with a `NamedTuple`:

```
    if (isNamedTupleClass(obj)) {
      // Maybe here or inside registerNamedTuple
      return registerNamedTuple(obj, loc);
    }
```

SplitInfinity reviewed (commented):
I annotated some things I want to discuss further.

gmagogsfm reviewed (commented):
Generally looks good. There are a few broken tests though with quite diverse error messages.

SplitInfinity commented on torch/_jit_internal.py:
```diff
@@ -216,6 +229,43 @@ def __getattr__(self, key):
+    #   for name, value in cls.__dict__.items():
+    #       if getattr(NamedTuple, name) is None and getattr(tuple, name) is None:
+    #           # Must be a user-defined function.
+    allowed_methods = [
```
It does - I have included the union of all the built-in types from 3.6 and 3.8.

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

[ghstack-poisoned]

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

[ghstack-poisoned]

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

[ghstack-poisoned]

gmagogsfm commented on torch/_jit_internal.py:
```diff
@@ -216,6 +229,55 @@ def __getattr__(self, key):
+    # Unfortunately, there are some NamedTuples already being used that do have methods that are never
+    # called. Allow these to compile for now.
+    can_compile_allowlist = {
+        torch.nn.utils.rnn.PackedSequence,
```
Are there more fb internal unscriptable classes than this one? If so, we can't use allowlist approach.

How about we annotate problematic classes instead of using allowlist? Some ideas:

1) Use @torch.jit.ignore on user-defined methods to make class scriptable
2) Add either a special member like `__torchscript_allow_user_defined_methods` on class

SplitInfinity replied:
I like the ignore approach, let's do that.

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

[ghstack-poisoned]

gmagogsfm reviewed (approved):
Nice!

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

Differential Revision: [D25490367](https://our.internmc.facebook.com/intern/diff/D25490367)

[ghstack-poisoned]

Commit:
Update on "[JIT] Display error message when NamedTuples have user-defined functions"


**Summary**
At present, `NamedTuples` are compiled by extracting the names and types
of their members only. `NamedTuples` can have user-defined methods and
static methods, but JIT completely ignores these, leading to confusion
when these methods are called in TorchScript.

This commit adds some code to detect when a `NamedTuple` has a
user-defined methods and static methods and throw an error.

**Test Plan**
This commit adds two test cases (one for user-defined methods, one for
static methods) to `test_jit_py3.py`.

**Fixes**
This commit fixes #48984.

Differential Revision: [D25490367](https://our.internmc.facebook.com/intern/diff/D25490367)

[ghstack-poisoned]

eellison commented:
I just want to note that this is a BC-incompatible change. It might break user models in the same way that it broke internal models.

SplitInfinity commented:
@eellison 

Given that this PR is just a stopgap for UX until we actually fix this, maybe adding this error message for temporary improved UX is not worth the trouble?

SplitInfinity commented:
Ping.

albanD commented on torch/nn/utils/rnn.py:
```diff
@@ -56,6 +56,7 @@ class PackedSequence(PackedSequence_):
         (i.e., they only pass in tensors conforming to this constraint).
 
     """
+    @ignore
```
Why do you need to ignore all of these functions? They were working before?
Also I don't see tests about this?
